mage_ai/server/websocket.py

```diff
@@ -40,25 +47,64 @@ def on_message(self, raw_message):
             manager = KernelManager(**connection)
             client = manager.client()
 
-            msg_id = client.execute(add_internal_output_info(code))
+            pipeline = Pipeline(pipeline_uuid, get_repo_path())
+            block = pipeline.get_block(block_uuid)
+            block_output = []
+            error = None
+            if block is not None and block.type in [BlockType.DATA_LOADER, BlockType.TRANSFORMER]:
+                try:
+                    output = asyncio.run(block.execute(custom_code=code))
+                    if len(output) > 0:
+                        for out in output:
+                            if type(out) == pd.DataFrame \
+                                and out.shape[0] > DATAFRAME_OUTPUT_SAMPLE_COUNT:
+
+                                out = out.iloc[:DATAFRAME_OUTPUT_SAMPLE_COUNT]
+                            block_output.append(out)
+                except:
+                    error = traceback.format_exc()
+                # Run with no code because we still need to send a message
+                msg_id = client.execute('')
+            else:
+                msg_id = client.execute(add_internal_output_info(code))
+
+            value = dict(
+                block_uuid=block_uuid,
+                block_output=block_output,
+                error=error,
+            )
 
-            WebSocketServer.running_executions_mapping.add((msg_id, block_uuid))
+            WebSocketServer.running_executions_mapping[msg_id] = value
         elif output:
             self.send_message(output)
 
     @classmethod
     def send_message(self, message: dict) -> None:
+        msg_type = message['msg_type']
         msg_id = message['msg_id']
-        msg_id_uuid_tuple = find(
-            lambda tup: tup[0] == msg_id,
-            list(WebSocketServer.running_executions_mapping),
+        msg_id_value = WebSocketServer.running_executions_mapping[msg_id]
+        uuid = msg_id_value['block_uuid']
+        output = msg_id_value['block_output']
+        error = msg_id_value['error']
+
+        output_dict = dict(uuid=uuid)
+        if error is not None:
+            output_dict['traceback'] = error
+        elif len(output) > 0:
+            df = find(lambda val: type(val) == pd.DataFrame, output)
+            if msg_type == 'execute_input':
```

Comment:
Is there a enum for msg_type?

Reply:
i don't think so, this msg_type is also set by the jupyter client which doesn't have an enum set